src/overlays/actors/ovl_player_actor/z_player.c

```diff
@@ -8071,47 +8089,80 @@ void Player_Action_808407CC(Player* this, PlayState* play) {
-        this->stateFlags2 &= ~PLAYER_STATE2_28;
+        (!(heathIsCritical = Health_IsCritical()) && ((this->idleType = (this->idleType + 1) & 1) != 0))) {
+        this->stateFlags2 &= ~PLAYER_STATE2_SPECIAL_IDLE;
         anim = Player_GetIdleAnim(this);
     } else {
-        this->stateFlags2 |= PLAYER_STATE2_28;
+        this->stateFlags2 |= PLAYER_STATE2_SPECIAL_IDLE;
+
         if (this->stateFlags1 & PLAYER_STATE1_CARRYING_ACTOR) {
+            // Normal idle animations will play if carrying an actor.
+            // Note that in this case, `PLAYER_STATE2_SPECIAL_IDLE` is still set even though the
+            // animation that plays isn't a special one.
             anim = Player_GetIdleAnim(this);
         } else {
-            sp38 = play->roomCtx.curRoom.behaviorType2;
+            // Pick special type based on room behavior.
+            // This may be changed below.
+            specialIdleType = play->roomCtx.curRoom.behaviorType2;
+
             if (heathIsCritical) {
-                if (this->unk_6AC >= 0) {
-                    sp38 = 7;
-                    this->unk_6AC = -1;
+                if (this->idleType >= PLAYER_IDLE_NORMAL) {
+                    specialIdleType = SPECIAL_IDLE_CRIT_HEALTH_START;
+
+                    // When health is critical, `idleType` will not be updated.
+                    // It will stay as `PLAYER_IDLE_CRIT_HEALTH` until health is no longer critical.
+                    this->idleType = PLAYER_IDLE_CRIT_HEALTH;
                 } else {
-                    sp38 = 8;
+                    // Keep looping the critical health animation until critical health ends
+                    specialIdleType = SPECIAL_IDLE_CRIT_HEALTH_LOOP;
                 }
             } else {
-                sp34 = Rand_ZeroOne() * 5.0f;
-                if (sp34 < 4) {
-                    if (((sp34 != 0) && (sp34 != 3)) || ((this->rightHandType == PLAYER_MODELTYPE_RH_SHIELD) &&
-                                                         ((sp34 == 3) || (Player_GetMeleeWeaponHeld2(this) != 0)))) {
-                        if ((sp34 == 0) && Player_HoldsTwoHandedWeapon(this)) {
-                            sp34 = 4;
+                commonType = Rand_ZeroOne() * 5.0f;
+
+                // There is a 4/6 chance that a common special type will be considered.
+                // However it may get rejected by the conditions below.
+                // The type determined by `curRoom.behaviorType2` will be used if a common type is rejected.
+                if (commonType <= COMMON_SPECIAL_IDLE(SPECIAL_IDLE_ADJUST_SHIELD)) {
+                    // `SPECIAL_IDLE_ADJUST_TUNIC` and `SPECIAL_IDLE_TAP_FEET` are accepted unconditionally.
+                    // The sword and shield related common types have extra restrictions.
+                    if (((commonType != COMMON_SPECIAL_IDLE(SPECIAL_IDLE_SWORD_SWING)) &&
+                         (commonType != COMMON_SPECIAL_IDLE(SPECIAL_IDLE_ADJUST_SHIELD))) ||
```

Comment:
isn't there some bug here where link will do something with his shield even though he doesn't have a shield? how does that work?

Reply:
I do know of that bug (very easy to see in game) but I do not know how it works. It clearly checks for right hand model type as "shield". Model type stuff is pretty confusing and havent looked into it too much, which is why I didnt comment

Reply:
I think PLAYER_MODELGROUP_SWORD_AND_SHIELD always sets PLAYER_MODELTYPE_RH_SHIELD regardless of whether you have a shield or not, so I think the bug here is assuming that PLAYER_MODELTYPE_RH_SHIELD means you have a shield in hand